### ISSUE TITLE
Does gotestsum support adding properties to test cases in the junit xml result file ? 

### ISSUE THREAD

**Issue by Victor-Chirinian-Vimeo** [gotestyourself/gotestsum#311]
Hello,
I was wondering if there was a way to add properties to a test case and be available in the xml result file.

`<testcase classname="tests.LoginTests" name="test_case_1" time="650">
      <properties>
        <property name="property1" value="value1"/>
     </properties>`

**Comment by dnephin** [comment]
Hello!

There is currently no way to do that with `gotestsum`, but I'd love to learn more about your use case.

What kind of data do you want to associate with the test case? Is it data that `gotestsum` may already have access to from the test output, or would it come from some other source? If the data is already available to `gotestsum`, it might be easy to get those properties added to the xml.

How will these properties be consumed? Is there another system that will read these properties? 

[XSLT](https://developer.mozilla.org/en-US/docs/Web/XSLT) is a tool that could be used to modify the junit.xml document and add those properties. I haven't used XSLT in many years, and never with Go, but I think it would be possible to write something using XSLT to add those properties after the file is created.

**Comment by Victor-Chirinian-Vimeo** [comment]
@dnephin Thanks for replying to my question. The data being added to the report would come from another source. We are currently integrating our Test Automation Suite results with TestRail by including a TestId in the name and parsing it out of the name attribute in the report. It would be nice to do something similar to this instead where the identifier is included in the report as a testcase property: 

`<testsuites name="test suites root">
  <testsuite failures="0" errors="0" skipped="1" tests="1" time="3049" name="tests.LoginTests">
    <testcase classname="tests.LoginTests" name="test_case_1" time="650">
     <properties>
         <property name="test_id" value="C123"/>
     </properties>
    </testcase>
  </testsuite>
</testsuites>`

**Comment by dnephin** [comment]
Thanks for the explanation! I'd be happy to explore some options for solving this problem. 

I think the few things we need to figure out are:
1. Where does the `test_id` value come from?
2. How does `gotestsum` store the association of test name (including package) to `test_id`? It might be nice to make this generic metadata associated with a test case instead of only `test_id`.
3. How would someone configure `gotestsum` with the xml tags and attributes to render the metadata.

I've got some initial thoughts, but I'm hoping we'll find other options too.

**Where does the `test_id` value come from?**

Two options that come to mind:
1. `gotestsum` could parse the output of tests and look for some kind of identifier that communicates the metadata (ex: `GOTESTSUM_METDATA=test_id=<value>`). This would require that every test case print the ID. Maybe it would be possible to print the full `xml` to be added to the report? In some ways this is similar to how github actions accepts values from commands.
2. `gotestsum` could read `key=value` pairs from a file, or maybe even `key=<xml>` pairs, where `key` is the package and test name. Maybe a JSON document would be a better option than `key=value` pairs.  If the file can contain the full xml to inject for each test case, that would solve problem 3.

What do you think of these options? Is there a better way to communicate this metadata?

I think the other two problems will largely depend on how we solve this first one.

**Comment by Victor-Chirinian-Vimeo** [comment]
1. The Test Id's are generated by [TestRail](https://www.testrail.com/), which is a test case management system. They are used to report results from automated tests using their API. I can imagine folks would/could use this or any other system to report test results to, if they are not running unit tests.
2. I agree, storing metadata for each test should be generic and not specific to test_id. 
3. Could we expose a package that allows the individual test to edit or add data to the generated JUNIT xml file ?  Is the XML file generated at the end of execution ? Ideally it would be great if we could add it in a test setup or teardown method.

**Comment by Victor-Chirinian-Vimeo** [comment]
@dnephin This sounds great, thanks for writing out the flow diagram. I think you are right, given what you have already mentioned, an external metadata file would be the best approach. Can we have the JunitXMLProperties be an array of json objects that are just  {"name": "value"}, instead of actually including the xml tags ?

**Comment by dnephin** [comment]
That's is definitely possible. I like the idea of accepting XML tags because I'm not sure what other systems might expect for the xml tags and attributes. Accepting XML would allow this feature to be used by other systems, but definitely does put more of a burden on the user to ensure they are producing valid XML.

Maybe the library that is responsible for creating this file could take on the responsibility of ensuring valid XML is written to the file? Or there other problems with this approach that I've missed?

**Comment by Victor-Chirinian-Vimeo** [comment]
No, I think this is great, and appreciate you taking the time to respond and comment on this enhancement. Please let me know if you need anything else from me. I would be happy to help test this enhancement when it becomes available.

**Comment by dnephin** [comment]
There is a Go proposal open (https://github.com/golang/go/issues/43936) to Allow Go Tests to Pass Metadata. If this is accepted this would be a great solution to the problem. The test suite could write the metadata, and `gotestsum` could read it and add it to the Junit.xml. All that would be needed is a convention about how to structure the metadata.

**Comment by MadddinTribleD** [comment]
Correct my but this already works out of the box now?
Use the `t.Attr(...)` mentioned here https://github.com/golang/go/issues/43936#issuecomment-2807744725

```bash
gotestsum --junitfile junit.file -- ./...
```

something_test.go
```go
func TestSomething(t *testing.T) {
	t.Run("My test", func(t *testing.T) {
		t.Attr("test_id", "C123")
	})
}
```

junit.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="2" failures="0" errors="0" time="0.526018">
	<testsuite tests="2" failures="0" time="0.526000" name="example" timestamp="2026-07-21T08:39:19+02:00">
		<properties>
			<property name="go.version" value="go1.26.0 darwin/arm64"></property>
		</properties>
		<testcase classname="example" name="TestSomething/My_test" time="0.000000">
			<properties>
				<property name="test_id" value="C123"></property>
			</properties>
		</testcase>
		<testcase classname="example" name="TestSomething" time="0.000000"></testcase>
	</testsuite>
</testsuites>
```
